scripts/kiwix-resources

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python3
+
+'''
+Copyright 2022 Veloman Yunkan <veloman.yunkan@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3 of the License, or any
+later version.
+
+This program is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+02110-1301, USA.
+'''
+
+import argparse
+import hashlib
+import os.path
+import re
+
+def read_resource_file(resource_file_path):
+    with open(resource_file_path, 'r') as f:
+        return [line.strip() for line in f]
+
+def list_resources(resource_file_path):
+    for resource_path in read_resource_file(resource_file_path):
+        print(resource_path)
+
+def get_resource_revision(base_dir, resource_path):
+    with open(os.path.join(base_dir, resource_path), 'rb') as f:
+        return hashlib.sha1(f.read()).hexdigest()[:8]
+
+resource_revisions = {}
+
+def fill_resource_revisions(resource_file_path):
+    base_dir = os.path.dirname(os.path.realpath(resource_file_path))
+    for resource in read_resource_file(resource_file_path):
+        resource_revisions[resource] = get_resource_revision(base_dir, resource)
+
+RESOURCE_WITH_CACHEID_URL_PATTERN=r'"((.*)/skin/([^"?]+))\?KIWIXCACHEID([^"]*)"'
+
+def set_cacheid(resource_matchobj):
+    path = resource_matchobj.group(1)
+    resource = 'skin/' + resource_matchobj.group(3)
+    extra_query = resource_matchobj.group(4)
+    cacheid = 'cacheid=' + resource_revisions[resource]
+    return '"' + path + '?' + cacheid + extra_query + '"'
+
+def preprocess_line(line):
+    if 'KIWIXCACHEID' in line:
+        line = re.sub(RESOURCE_WITH_CACHEID_URL_PATTERN, set_cacheid, line)
+        assert not 'KIWIXCACHEID' in line
+    return line
```

Comment:
I think this could be better rewriten with something link:

```python
RESOURCE_WITH_CACHEID_URL_REG = re.compile(r'(?P<pre_url>".*/(?P<resource>skin/[^"?]+)\?.*)KIWIXCACHEID(?P<post_url>[^"]*")'

def set_cacheid(resource_matchobj):
    resource = resource_matchopj.group('resource')
    cacheid = 'cacheid='+resource_revisions[resource]
    return resource_matchobj.group('pre_url') + cacheid + resource_matchobj.group('post_url')

def process_line(line):
    if 'KIWIXCACHEID' in line:
        RESOURCE_WITH_CACHEID_URL_REG.sub(set_cacheid, line)
        assert not 'KIWIXCACHEID' in line
    return line
```
(not tested)

Reply:
Done but in a way that still requires `KIWIXCACHEID` to immediately follow the `?` symbol. This is done for safety and doesn't create any obstacles.